src/components/ActionSheetAwareScrollView/ActionSheetKeyboardSpace.tsx

```diff
@@ -0,0 +1,411 @@
+    useAnimatedStyle,
+    useDerivedValue,
+    withSpring,
+    useAnimatedReaction,
+    runOnJS,
+    withSequence,
+    withTiming,
+    useSharedValue,
+    interpolate,
+} from 'react-native-reanimated';
+
+import useSafeAreaInsets from '@hooks/useSafeAreaInsets';
+import useThemeStyles from '@hooks/useThemeStyles';
+
+import { useKeyboardHandler } from 'react-native-keyboard-controller';
+import useWindowDimensions from '@hooks/useWindowDimensions';
+import useStyleUtils from '@hooks/useStyleUtils';
+
+import {Actions, States, ActionSheetAwareScrollViewContext} from './ActionSheetAwareScrollViewContext';
+
+const KeyboardState = {
+    UNKNOWN: 0,
+    OPENING: 1,
+    OPEN: 2,
+    CLOSING: 3,
+    CLOSED: 4
+}
+const useAnimatedKeyboard = () => {
+    const state = useSharedValue(KeyboardState.UNKNOWN);
+    const height = useSharedValue(0);
+    const progress = useSharedValue(0);
+    const heightWhenOpened = useSharedValue(0);
+
+    useKeyboardHandler({
+        onStart: (e) => {
+            "worklet";
+
+            // save the last keyboard height
+            if (e.height === 0) {
+                heightWhenOpened.value = height.value;
+            }
+
+            // console.log("onStart", e, new Date().getTime());
+
+            if (e.height > 0) {
+                state.value = KeyboardState.OPENING;
+            } else {
+                state.value = KeyboardState.CLOSING;
+            }
+        },
+        onMove: (e) => {
+            "worklet";
+
+            // console.log("onMove", e, new Date().getTime());
+
+            progress.value = e.progress;
+            height.value = e.height;
+        },
+        onEnd: (e) => {
+            "worklet";
+
+            // console.log("onEnd", e, new Date().getTime());
+
+            if (e.height > 0) {
+                state.value = KeyboardState.OPEN;
+            } else {
+                state.value = KeyboardState.CLOSED;
+            }
+
+            height.value = e.height;
+            progress.value = e.progress;
+        },
+    }, []);
+
+    return { state, height, heightWhenOpened, progress };
+};
+const setInitialValueAndRunAnimation = (value: number, animation: number) => {
+    "worklet";
+
+    return withSequence(
+        withTiming(value, { duration: 0 }),
+        animation,
+    )
+}
+
+const useSafeAreaPaddings = () => {
+    const StyleUtils = useStyleUtils();
+    const insets = useSafeAreaInsets();
+    const {paddingTop, paddingBottom} = StyleUtils.getSafeAreaPadding(insets ?? undefined);
+
+    return { top: paddingTop, bottom: paddingBottom };
+};
+
+const config = {
+    mass: 3,
+    stiffness: 1000,
+    damping: 500,
+};
+
+function ActionSheetKeyboardSpace(props: ViewProps) {
+    const styles = useThemeStyles();
+    const safeArea = useSafeAreaPaddings();
+    const keyboard = useAnimatedKeyboard();
+
+    // similar to using `global` in worklet but it's just a local object
+    const [syncLocalWorkletState] = useState({
```

Comment:
Why would we need to use `useState` if we're never modifying the state?

Reply:
Seems like this should use `useRef` instead?

Reply:
@roryabraham originally it was created by @terrysahaidak so I think he can provide more output.

But as I can understand such construction serves as an analog of `useRef` in reanimated world. If we re-assign a shared value, then corresponding `useDerivedValue`/`useAnimatedReaction` effects will be run. And in order to save a value somewhere and don't re-run reanimated lifecycles we're using state variable with direct assigning (on next reanimated hooks execution they will capture a new variable into the closure).

Such trick will not work with `useRef` I guess, because `ref` will always point to the same object (it has only `current` property that can be modified).

@terrysahaidak also don't hesitate to jump into conversation if my explanation is not correct 👀

Reply:
I think it can also be something like `const syncLocalWorkletState = useRef({}).current`. It's still a bit unclear for me if this gonna work with useRef the same way it would work with useState which should preserve its state in React's Concurrent mode. useState was the saver bet.